src/codecs/rotate/processor.ts

```diff
@@ -1,73 +1,30 @@
+import { RotateOptions, RotateModuleInstance } from './processor-meta';
 
-export function rotate(data: ImageData, opts: RotateOptions): ImageData {
-  const { rotate } = opts;
-  const flipDimensions = rotate % 180 !== 0;
-  const { width: inputWidth, height: inputHeight } = data;
-  const outputWidth = flipDimensions ? inputHeight : inputWidth;
-  const outputHeight = flipDimensions ? inputWidth : inputHeight;
-  const out = new ImageData(outputWidth, outputHeight);
-  let i = 0;
+export async function rotate(
+  data: ImageData,
+  opts: RotateOptions,
+): Promise<ImageData> {
+  const flipDimensions = opts.rotate % 180 !== 0;
+  // Number of wasm memory pages (á 64KiB) needed to store the image twice.
+  const bytesPerImage = data.width * data.height * 4;
+  const numPagesNeeded = Math.ceil(bytesPerImage * 2 / (64 * 1024));
+  // const memory = new Uint8ClampedArray(data.data.length * 2);
+  // const memory = new WebAssembly.Memory({
+  //   initial: Math.ceil(data.width * data.height * 4 * 2 / (64 * 1024)),
+  // });
+  const { instance } = (await (WebAssembly as any).instantiateStreaming(
+    fetch(wasmUrl),
+  )) as { instance: RotateModuleInstance };
 
-  // In the straight-copy case, d1 is x, d2 is y.
-  // x starts at 0 and increases.
-  // y starts at 0 and increases.
-  let d1Start = 0;
-  let d1Limit = inputWidth;
-  let d1Advance = 1;
-  let d1Multiplier = 1;
-  let d2Start = 0;
-  let d2Limit = inputHeight;
-  let d2Advance = 1;
-  let d2Multiplier = inputWidth;
-
-  if (rotate === 90) {
-    // d1 is y, d2 is x.
-    // y starts at its max value and decreases.
-    // x starts at 0 and increases.
-    d1Start = inputHeight - 1;
-    d1Limit = inputHeight;
-    d1Advance = -1;
-    d1Multiplier = inputWidth;
-    d2Start = 0;
-    d2Limit = inputWidth;
-    d2Advance = 1;
-    d2Multiplier = 1;
-  } else if (rotate === 180) {
-    // d1 is x, d2 is y.
-    // x starts at its max and decreases.
-    // y starts at its max and decreases.
-    d1Start = inputWidth - 1;
-    d1Limit = inputWidth;
-    d1Advance = -1;
-    d1Multiplier = 1;
-    d2Start = inputHeight - 1;
-    d2Limit = inputHeight;
-    d2Advance = -1;
-    d2Multiplier = inputWidth;
-  } else if (rotate === 270) {
-    // d1 is y, d2 is x.
-    // y starts at 0 and increases.
-    // x starts at its max and decreases.
-    d1Start = 0;
-    d1Limit = inputHeight;
-    d1Advance = 1;
-    d1Multiplier = inputWidth;
-    d2Start = inputWidth - 1;
-    d2Limit = inputWidth;
-    d2Advance = -1;
-    d2Multiplier = 1;
-  }
-
-  const inB = new Uint32Array(data.data.buffer);
-  const outB = new Uint32Array(out.data.buffer);
-  for (let d2 = d2Start; d2 >= 0 && d2 < d2Limit; d2 += d2Advance) {
-    for (let d1 = d1Start; d1 >= 0 && d1 < d1Limit; d1 += d1Advance) {
-      const start = ((d1 * d1Multiplier) + (d2 * d2Multiplier));
-      outB[i] = inB[start];
-      i += 1;
-    }
-  }
-
-  return out;
+  instance.exports.memory.grow(numPagesNeeded);
```

Comment:
I guess there isn't a way to set the memory during initialisation?

Reply:
Haven’t found a way. @MaxGraey?

Reply:
you can do it during instantiation inside `env` object:
```js
const limitInBytes = 2 * 1024 * 1024; // 2 Mb
const limitPages = ((limitInBytes + 0xffff) & ~0xffff) >>> 16; // convert to pages
env: {
   ...
   memory: new WebAssembly.Memory({ initial: 1, maximum: limitPages })
} 
```

Also need add `--importMemory` for `asc`

Reply:
Ah cool. Thanks!

Reply:
I was missing the flag.